android/src/app/organicmaps/maplayer/MapButtonsController.java

```diff
@@ -301,26 +308,32 @@ private int getViewTopOffset(float translation, View v)
     return (int) (translation + v.getTop());
   }
 
+  @Override
+  public void onStop()
+  {
+    super.onStop();
+    mPlacePageViewModel.getPlacePageDistanceToTop().removeObserver(mPlacePageDistanceToTopObserver);
+    mMapButtonsViewModel.getButtonsHidden().removeObserver(mButtonHiddenObserver);
+    mMapButtonsViewModel.getMyPositionMode().removeObserver(mMyPositionModeObserver);
+    mMapButtonsViewModel.getMapLayerMode().removeObserver(mMapLayerModeObserver);
+    mMapButtonsViewModel.getSearchOption().removeObserver(mSearchOptionObserver);
+  }
   public void onResume()
   {
     super.onResume();
+    mPlacePageViewModel.getPlacePageDistanceToTop().observe(requireActivity(), mPlacePageDistanceToTopObserver);
+    mMapButtonsViewModel.getButtonsHidden().observe(requireActivity(), mButtonHiddenObserver);
+    mMapButtonsViewModel.getMyPositionMode().observe(requireActivity(), mMyPositionModeObserver);
+    mMapButtonsViewModel.getMapLayerMode().observe(requireActivity(), mMapLayerModeObserver);
+    mMapButtonsViewModel.getSearchOption().observe(requireActivity(), mSearchOptionObserver);
```

Comment:
1. What are these different observers, and what is their role/usefulness for the user, especially when some dialogs are displayed on top of the paused activity?
2. What is the full list of dialogs that may pause the activity? E.g. the "location is disabled" dialog (it means that subscribing to location updates doesn't make sense yet, right?), what else?

Reply:
The observers related to `mMapButtonsViewModel` allow the buttons to react to changes in the app state, such as when the position mode or the layers change, etc.

The observer on `mPlacePageViewModel.getPlacePageDistanceToTop` allows the buttons to follow/avoid the place page when it slides. This observer cannot be removed in `onPause` otherwise the buttons will not follow the place page if it is sliding while there is a popup over the screen (eg you click share and swipe down the place page at the same time, the place page will slide while the share popup pauses the activity). 

To make the code easier to understand, I chose to remove every observer in `onStop`.

Reply:
l. That's a leak. Resume may be called many times.
2. It would be safer to leave the old behavior, and only subscribe/unsubscribe in onStop the one needed observer. It will also save users batteries in many cases.

Reply:
For example `onResume` may be called right after `onPause` without triggering `onStop`. That makes handling lifecycle more complex.

Reply:
It is an inconsistent pattern to start something in onResume but stop it onStop.

A more general question: the ViewModel approach doesn't make the code easier, it's more connected now than before the refactoring. Are there any better ways? Maybe not following official "pattern recommendations", but something simpler, with fewer dependencies and fewer lines of code?

Reply:
I don't see why it is inconsistent as onResume is called both on activity start and on activity resume, but I don't have that much android experience.

After reading the docs a bit more, I think we may be able to simplify this and only observe/remove in onCreate/onDestroy because the viewmodels are linked to the activity lifecycle and thus will never fire when the app goes into background.

Regarding your question, it does make the code simpler. You now have a clear dependency on only the viewmodels, while before you had a mess of callbacks from different classes interacting with each other. Fragments should not depend on any other class than viewmodels. It makes it a lot easier to handle lifecycle as you don't have to worry about using omSaveState, the viewmodel does it for you.

Reply:
nit: I'd prefer additional var: ```final Activity activity = requireActivity();```